src/client/components/vision2/camera/line_projection.ts

```diff
@@ -0,0 +1,105 @@
+import * as THREE from 'three'
+
+import { Vector2 } from '../../../math/vector2'
+import { Vector3 } from '../../../math/vector3'
+import { Vector4 } from '../../../math/vector4'
+import { mesh } from '../../three/builders'
+import { planeGeometry } from '../../three/builders'
+import { rawShader } from '../../three/builders'
+import { shaderMaterial } from '../../three/builders'
+import { Canvas } from '../../three/three'
+
+import { Lens } from './model'
+import fragmentShader from './shaders/line_projection.frag'
+import vertexShader from './shaders/line_projection.vert'
+
+export interface ConeSegment {
+  /** The camera space central axis of the cone. Only needed for a plane if the start and end are (anti-)parallel. */
+  axis: Vector3,
+  /** The camera space vector pointing to the start of the segment. */
+  start: Vector3,
+  /** The camera space vector pointing to the end of the segment. */
+  end: Vector3,
+  /** The colour of the line to draw. */
+  color?: Vector4,
+  /** The width of the line to draw on the screen in pixels. */
+  lineWidth?: number
+}
+
+export class LineProjection {
+  constructor(private readonly canvas: Canvas, private readonly lens: Lens) {
+  }
+
+  static of(canvas: Canvas, lens: Lens) {
+    return new LineProjection(canvas, lens)
+  }
+
+  /** Draws a plane projected to infinity in world space. */
+  plane({ axis, color, lineWidth }: { axis: Vector3, color?: Vector4, lineWidth?: number }) {
+    // Pick an arbitrary orthogonal vector
+    const start = Vector3.fromThree(
+      (!axis.x && !axis.y)
+        ? new THREE.Vector3(0, 1, 0)
+        : new THREE.Vector3(-axis.y, axis.x, 0).normalize(),
+    )
+    return this.coneSegment({ axis, start, end: start, color, lineWidth })
+  }
+
+  /** Draws a segment of a plane projected to infinity in world space. */
+  planeSegment(segment: { axis?: Vector3, start: Vector3, end: Vector3, color?: Vector4, lineWidth?: number }) {
+    return this.coneSegment({
+      ...segment,
+      axis: Vector3.fromThree(
```

Comment:
Still need to use the `segment.axis` here if provided. The problem is that when you have a 360 or 180-degree angle, crossing them gives no normal vector to project the plane on. So either we need to find a way around that or provide an axis here.